File: breakpilot-idea/src/main/kotlin/bridge/BridgeClient.kt
package bridge

import com.google.gson.Gson
import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.project.Project
import java.io.File
import java.net.URI
import java.net.http.HttpClient
import java.net.http.WebSocket
import java.time.Duration
import java.time.Instant
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.ConcurrentHashMap

class BridgeClient(private val project: Project) : Disposable {
    private val defaultBridgeUrl = "ws://127.0.0.1:57987/bridge"
    private val gson = Gson()
    private val listeners = mutableListOf<(BridgeMessage) -> Unit>()
    private val scheduler: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    private val disposed = AtomicBoolean(false)
    @Volatile private var socket: WebSocket? = null
    private val connectionLock = Any()
    private val connectionGeneration = AtomicLong(0)
    private var explicitBridgeUrl: String? = null
    private var currentBridgeUrl: String? = null
    private var currentInstanceId: String? = null
    private var heartbeat: ScheduledFuture<*>? = null
    private val pending = mutableListOf<BridgeMessage>()
    private val requestGenerations = ConcurrentHashMap<String, Long>()

    fun connect(url: String? = null) {
        if (url != null) explicitBridgeUrl = url
        if (disposed.get()) return
        val target = resolveBridgeTarget() ?: run {
            closeSocket()
            scheduleReconnect()
            return
        }
        if (socket != null && currentBridgeUrl == target.url && currentInstanceId == target.instanceId) return
        val oldSocket: WebSocket?
        val generation = synchronized(connectionLock) {
            val next = connectionGeneration.incrementAndGet()
            oldSocket = socket
            socket = null
            next
        }
        requestGenerations.clear()
        synchronized(pending) { pending.clear() }
        oldSocket?.abort()
        currentBridgeUrl = target.url
        currentInstanceId = target.instanceId
        val client = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(3))
            .build()
        client.newWebSocketBuilder()
            .connectTimeout(Duration.ofSeconds(3))
            .buildAsync(URI.create(target.url), Listener(generation))
            .exceptionally {
                if (connectionGeneration.get() == generation) scheduleReconnect()
                null
            }
    }

    fun onMessage(listener: (BridgeMessage) -> Unit) {
        listeners += listener
    }

    fun send(message: BridgeMessage) {
        enqueueOrSend(message)
    }

    private fun enqueueOrSend(message: BridgeMessage) {
        synchronized(connectionLock) {
            val correlationId = message.requestId ?: message.confirmationId
            val generation = connectionGeneration.get()
            if (correlationId != null && requestGenerations[correlationId] != generation) return
            val current = socket
            if (current == null) {
                synchronized(pending) { pending += message }
                return
            }
            val outbound = message.copy(timestamp = Instant.now().toString())
            current.sendText(gson.toJson(outbound), true)
            if (correlationId != null) requestGenerations.remove(correlationId)
        }
    }

    private fun sendRegister() {
        enqueueOrSend(
            BridgeMessage(
                type = MessageTypes.IdeRegister,
                ide = "idea",
                workspaceRoot = project.basePath,
                debuggerProtocolVersion = 2,
                debuggerFeatures = mapOf(
                    "breakpointUpdate" to true,
                    "eventStream" to true,
                    "stackPagination" to true,
                    "variableHandles" to true,
                    "nativeSetVariable" to true,
                    "causalDebugStart" to true
                ),
                capabilities = mapOf(
                    "visualBreakpoints" to true,
                    "debugCommands" to true,
                    "confirmationDialog" to true,
                    "structuredConfirmation" to true,
                    "consentSettings" to true,
                    "toolWindow" to true,
                    "variableSnapshot" to true,
                    "threads" to true,
                    "stackTrace" to true,
                    "runToLine" to true,
                    "setVariable" to true,
                    "setVariableMode" to "evaluateAssignment",
                    "provider" to "xdebugger"
                )
            )
        )
    }

    private fun sendHeartbeat() {
        if (socket != null) enqueueOrSend(BridgeMessage(type = MessageTypes.IdeHeartbeat))
    }

    private fun scheduleHeartbeat() {
        heartbeat?.cancel(false)
        heartbeat = scheduler.scheduleAtFixedRate({
            if (!disposed.get()) sendHeartbeat()
        }, 5, 5, TimeUnit.SECONDS)
    }

    private fun scheduleReconnect() {
        if (disposed.get()) return
        scheduler.schedule({
            if (!disposed.get()) connect()
        }, 2, TimeUnit.SECONDS)
    }

    private fun resolveBridgeTarget(): BridgeTarget? {
        explicitBridgeUrl?.trim()?.takeIf { it.isNotEmpty() }?.let { return BridgeTarget(it, null) }
        return BridgeTarget(defaultBridgeUrl, null)
    }

    override fun dispose() {
        disposed.set(true)
        connectionGeneration.incrementAndGet()
        requestGenerations.clear()
        heartbeat?.cancel(false)
        socket?.sendClose(WebSocket.NORMAL_CLOSURE, "disposed")
        scheduler.shutdownNow()
    }

    private fun closeSocket() {
        connectionGeneration.incrementAndGet()
        requestGenerations.clear()
        heartbeat?.cancel(false)
        heartbeat = null
        socket?.abort()
        socket = null
        synchronized(pending) { pending.clear() }
    }

    private fun emitLocal(message: BridgeMessage) {
        ApplicationManager.getApplication().invokeLater {
            listeners.forEach { it(message) }
        }
    }

    private inner class Listener(private val generation: Long) : WebSocket.Listener {
        private val incoming = StringBuilder()

        override fun onOpen(webSocket: WebSocket) {
            val accepted = synchronized(connectionLock) {
                if (disposed.get() || connectionGeneration.get() != generation) false
                else {
                    socket = webSocket
                    true
                }
            }
            if (!accepted) {
                webSocket.abort()
                return
            }
            webSocket.request(1)
            synchronized(pending) { pending.clear() }
            sendRegister()
            emitLocal(BridgeMessage(type = MessageTypes.BridgeConnected, workspaceRoot = project.basePath))
            scheduleHeartbeat()
        }

        override fun onText(webSocket: WebSocket, data: CharSequence, last: Boolean): CompletionStage<*> {
            if (!isCurrent(webSocket, generation)) {
                webSocket.request(1)
                return CompletableFuture.completedFuture(null)
            }
            incoming.append(data)
            if (last) {
                val text = incoming.toString()
                incoming.setLength(0)
                val message = gson.fromJson(text, BridgeMessage::class.java)
                if (message.type == MessageTypes.BridgeWelcome && message.workspaceRoot != null && project.basePath != null) {
                    if (File(message.workspaceRoot).absolutePath != File(project.basePath!!).absolutePath) {
                        webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "workspace mismatch")
                        return CompletableFuture.completedFuture(null)
                    }
                }
                if (message.type == MessageTypes.BridgeRejected) {
                    webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "bridge rejected")
                    return CompletableFuture.completedFuture(null)
                }
                val correlationId = message.requestId ?: message.confirmationId
                if (correlationId != null) requestGenerations[correlationId] = generation
                ApplicationManager.getApplication().invokeLater {
                    if (!isCurrent(webSocket, generation)) return@invokeLater
                    listeners.forEach { it(message) }
                }
            }
            webSocket.request(1)
            return CompletableFuture.completedFuture(null)
        }

        override fun onClose(webSocket: WebSocket, statusCode: Int, reason: String): CompletionStage<*> {
            if (clearCurrent(webSocket, generation)) {
                synchronized(pending) { pending.clear() }
                requestGenerations.clear()
                heartbeat?.cancel(false)
                emitLocal(BridgeMessage(type = MessageTypes.BridgeDisconnected, workspaceRoot = project.basePath))
                scheduleReconnect()
            }
            return CompletableFuture.completedFuture(null)
        }

        override fun onError(webSocket: WebSocket, error: Throwable) {
            if (clearCurrent(webSocket, generation)) {
                synchronized(pending) { pending.clear() }
                requestGenerations.clear()
                heartbeat?.cancel(false)
                emitLocal(BridgeMessage(type = MessageTypes.BridgeDisconnected, workspaceRoot = project.basePath))
                scheduleReconnect()
            }
        }
    }

    private fun isCurrent(webSocket: WebSocket, generation: Long): Boolean = synchronized(connectionLock) {
        socket === webSocket && connectionGeneration.get() == generation && !disposed.get()
    }

    private fun clearCurrent(webSocket: WebSocket, generation: Long): Boolean = synchronized(connectionLock) {
        if (socket !== webSocket || connectionGeneration.get() != generation) false
        else {
            socket = null
            true
        }
    }

    private data class BridgeTarget(val url: String, val instanceId: String?)
}

File: breakpilot-vscode/src/bridge/BridgeClient.ts
import * as vscode from "vscode";
import * as fs from "node:fs";
import * as path from "node:path";
import { AnyRecord, BridgeMessage, MessageTypes } from "./MessageProtocol";

export type BridgeConnectionState = "disconnected" | "connecting" | "connected" | "rejected";

export class BridgeClient {
  private socket?: WebSocket;
  private heartbeat?: NodeJS.Timeout;
  private reconnect?: NodeJS.Timeout;
  private watcher?: vscode.FileSystemWatcher;
  private currentUrl?: string;
  private currentInstanceId?: string;
  private explicitBridgeUrl?: string;
  private state: BridgeConnectionState = "disconnected";
  private disposed = false;
  private listeners = new Set<(message: BridgeMessage) => void>();
  private pending: BridgeMessage[] = [];
  private readonly connectionEmitter = new vscode.EventEmitter<BridgeConnectionState>();

  readonly onDidChangeConnectionState = this.connectionEmitter.event;

  constructor(private readonly context: vscode.ExtensionContext) {
    this.startManifestWatcher();
    this.context.subscriptions.push(
      vscode.workspace.onDidChangeWorkspaceFolders(() => {
        this.restartManifestWatcher();
        this.connect();
      })
    );
  }

  connect(url?: string) {
    if (this.disposed) return;
    if (url?.trim()) this.explicitBridgeUrl = url.trim();
    const target = this.resolveBridgeTarget();
    if (!target) {
      this.closeSocket();
      this.scheduleReconnect();
      return;
    }
    if (this.socket && this.currentUrl === target.url && this.currentInstanceId === target.instanceId) return;
    this.setState("connecting");
    this.currentUrl = target.url;
    this.currentInstanceId = target.instanceId;
    this.socket?.close();
    this.socket = new WebSocket(target.url);
    this.socket.onopen = () => {
      if (this.reconnect) clearTimeout(this.reconnect);
      this.setState("connected");
      this.send({
        type: MessageTypes.IdeRegister,
        ide: "vscode",
        workspaceRoot: this.workspaceRoot(),
        capabilities: this.capabilities(),
        debuggerProtocolVersion: 2,
        debuggerFeatures: this.debuggerFeatures()
      });
      this.flushPending();
      this.heartbeat = setInterval(() => {
        this.send({ type: MessageTypes.IdeHeartbeat });
      }, 5000);
    };
    this.socket.onmessage = (event) => {
      const message = this.parseMessage(event.data);
      if (!message) return;
      if (message.type === MessageTypes.BridgeWelcome && !this.workspaceMatches(message.workspaceRoot)) {
        this.socket?.close();
        return;
      }
      if (message.type === MessageTypes.BridgeRejected) {
        this.setState("rejected");
        this.socket?.close();
        return;
      }
      for (const listener of this.listeners) listener(message);
    };
    this.socket.onclose = () => {
      if (this.heartbeat) clearInterval(this.heartbeat);
      this.socket = undefined;
      this.setState("disconnected");
      this.scheduleReconnect();
    };
    this.socket.onerror = () => {
      this.socket = undefined;
      this.setState("disconnected");
      this.scheduleReconnect();
    };
  }

  onMessage(listener: (message: BridgeMessage) => void) {
    this.listeners.add(listener);
    return new vscode.Disposable(() => this.listeners.delete(listener));
  }

  send(message: BridgeMessage) {
    if (this.socket?.readyState === WebSocket.OPEN) {
      this.socket.send(JSON.stringify({ ...message, timestamp: new Date().toISOString() }));
      return;
    }
    if (message.type !== MessageTypes.IdeHeartbeat) {
      this.pending.push(message);
      if (this.pending.length > 200) this.pending.splice(0, this.pending.length - 200);
    }
  }

  dispose() {
    this.disposed = true;
    if (this.heartbeat) clearInterval(this.heartbeat);
    if (this.reconnect) clearTimeout(this.reconnect);
    this.watcher?.dispose();
    this.socket?.close();
    this.connectionEmitter.dispose();
  }

  capabilities(): AnyRecord {
    return {
      visualBreakpoints: true,
      debugCommands: true,
      confirmationDialog: true,
      structuredConfirmation: true,
      consentSettings: true,
      webviewPanel: true,
      variableSnapshot: true,
      runToLine: true,
      setVariable: true,
      setVariableMode: "evaluateAssignment",
      conditionalBreakpoints: true,
      hitConditionalBreakpoints: true,
      tracepoints: true,
      adoptSession: true,
      debugSessionTracking: true,
      breakpointHitTracking: true,
      evaluate: true,
      provider: "vscode-debug-api"
    };
  }

  debuggerFeatures(): Record<string, boolean> {
    return {
      breakpointUpdate: true,
      eventStream: true,
      stackPagination: true,
      variableHandles: true,
      nativeSetVariable: true,
      causalDebugStart: true
    };
  }

  workspaceRoot(): string | undefined {
    return vscode.workspace.workspaceFolders?.[0]?.uri.fsPath;
  }

  private resolveBridgeTarget(): { url: string; instanceId?: string } | null {
    if (this.explicitBridgeUrl) return { url: this.explicitBridgeUrl };
    const config = vscode.workspace.getConfiguration("breakpilot");
    const inspected = config.inspect<string>("bridgeUrl");
    const configured = inspected?.workspaceValue ?? inspected?.globalValue;
    if (configured && configured.trim()) return { url: configured.trim() };
    const workspaceRoot = this.workspaceRoot();
    if (workspaceRoot) {
      const bridgeFile = path.join(workspaceRoot, ".breakpilot", "bridge.json");
      try {
        const manifest = JSON.parse(fs.readFileSync(bridgeFile, "utf8")) as { bridgeUrl?: string; instanceId?: string };
        if (manifest.bridgeUrl) return { url: manifest.bridgeUrl, instanceId: manifest.instanceId };
      } catch {
        return null;
      }
    }
    return null;
  }

  private startManifestWatcher() {
    const workspaceRoot = this.workspaceRoot();
    if (!workspaceRoot) return;
    const pattern = new vscode.RelativePattern(workspaceRoot, ".breakpilot/bridge.json");
    this.watcher = vscode.workspace.createFileSystemWatcher(pattern);
    this.watcher.onDidCreate(() => this.connect());
    this.watcher.onDidChange(() => this.connect());
    this.watcher.onDidDelete(() => {
      this.currentUrl = undefined;
      this.currentInstanceId = undefined;
      this.closeSocket();
      this.emitLocal({ type: "bridge_disconnected", workspaceRoot });
      this.scheduleReconnect();
    });
    this.context.subscriptions.push(this.watcher);
  }

  private restartManifestWatcher() {
    this.watcher?.dispose();
    this.watcher = undefined;
    this.startManifestWatcher();
  }

  private closeSocket() {
    if (this.heartbeat) clearInterval(this.heartbeat);
    this.heartbeat = undefined;
    this.socket?.close();
    this.socket = undefined;
    this.setState("disconnected");
  }

  private scheduleReconnect() {
    if (this.disposed) return;
    if (this.reconnect) clearTimeout(this.reconnect);
    this.reconnect = setTimeout(() => this.connect(), 2000);
  }

  private flushPending() {
    const queued = this.pending.splice(0);
    for (const message of queued) this.send(message);
  }

  private parseMessage(data: unknown): BridgeMessage | null {
    try {
      return JSON.parse(String(data)) as BridgeMessage;
    } catch {
      return null;
    }
  }

  private workspaceMatches(remote: unknown): boolean {
    if (typeof remote !== "string" || !remote) return true;
    const local = this.workspaceRoot();
    if (!local) return false;
    return path.resolve(remote) === path.resolve(local);
  }

  private setState(state: BridgeConnectionState) {
    if (this.state === state) return;
    this.state = state;
    this.connectionEmitter.fire(state);
    if (state === "connected" || state === "disconnected" || state === "rejected") {
      this.emitLocal({
        type: state === "connected" ? MessageTypes.IdeRegistered : MessageTypes.BridgeDisconnected,
        workspaceRoot: this.workspaceRoot(),
        state
      });
    }
  }

  private emitLocal(message: BridgeMessage) {
    for (const listener of this.listeners) listener(message);
  }
}
